VSR/Correctness: Preserve Canonical Headers

sentientwaffle opened the pull request:
(Includes https://github.com/tigerbeetledb/tigerbeetle/pull/359, to ensure `repair()` commits enough ops that the pipeline can't run out of space when the new primary has to play catch-up.)

Fixes 2 correctness bugs:
1. Don't Truncate Canonical DVC
2. Bump Primary view_normal Early

## Don't Truncate Canonical DVC

### Summary

Truncating a canonical DVC's headers can fork the log.

### VOPR Example

Consider a cluster with:

- Replica 2 is leading a view change (view=17).
- Replica 1 is participating in the view=17 view change.
- Replica 0 is partitioned.
- `pipeline_prepare_queue_max = 6`

The DVC quorum:

    2: on_do_view_change: replica=1 view_normal=15 op=38 commit_min=31
    2: on_do_view_change: replica=2 view_normal=14 op=33 commit_min=28

primary_set_log_from_do_view_change_messages():

  1. Replica 2's `op_canonical` is 28 (computed by `primary_op_canonical_max()`).
  3. Replica 1's DVC included headers 33,34,35,36,37,38. These were loaded into Replica 2's headers since Replica 1 has the canonical view.
  4. Near the end of `primary_set_log_from_do_view_change_messages()`, `do_view_change_op_max()` finds a hash break between op 32/33, and discards 33…38, forking the log.

It discarded the entire (canonical!) hash chain from replica 1! (In this case those ops were uncommitted, but that doesn't necessarily need to be true.) In any case, by discarding those ops we revert to our old (incorrect) op 29, which otherwise would have been repaired. We go on to commit the wrong op 29.

According to `do_view_change_op_max()`, the rationale for discarding those breaks (i.e. breaks which follow the primary's `op_canonical`) is:

    /// If there is a hash chain break, none of the headers from the canonical DVCs replaced
    /// the broken (leftover uncanonical) op.
    /// Removing these is necessary for correctness and liveness, to ensure that
    /// disconnected headers do not remain in place in lieu of gaps.

As the first sentence hints, passing more headers in the DVC would decrease the probability of this error, but that isn't an actual fix. (Aside: This bug was likely hidden until recently because it is becomes less probably for larger pipelines, and I have recently been heavily testing small pipelines.)

### Simplified Example

With `pipeline_prepare_queue_max=3`:

```
  replica  headers                    commit_min  view_normal
        0  (1a 2a 3a 4a 5a 6a 7a 8a)          8a            4 (partitioned)
        1   1a 2a 3a 4a 5a{6a 7a 8a}          5a            4
        2   1a 2a 3a 4b                       3a            2
```

Prior to the do_view_change_op_max() truncation, replica's 2 headers are:

```
        2   1a 2a 3a 4b 5a 6a 7a 8a
                      ↑ hash break
```

Per CTRL, we can only discard headers that are *definitely uncommitted*.
A gap/break in replica 1's 6a,7a,8a would be definitively uncommitted and this safe to discard.
A break before _any_ of the view_normal_canonical's DVC headers though _must_ be repaired.

## Bump Primary view_normal Early

### Background

The new primary picks (in `primary_set_log_from_do_view_change_messages()`) at least one "canonical view". This is the highest `view_normal` from any DVC in its quorum.
The canonical DVC(s) headers are called the view's "canonical headers". These headers help select the new log head.

If multiple replicas have the same `view_normal`, there may be multiple canonical DVCs.
However, canonical headers _must_ be unambiguous. More specifically:

  - Any particular DVC may have gaps (1a, 2a, 4a), but no hash breaks (1a, 2a, 4b).
  - DVCs with the same `view_normal` _must_ not conflict.

### Bug

*DVCs with the same `view_normal` _must_ not conflict.*

Currently they can conflict, though. The heart of this bug is that the `view_normal` updates at a different time than the headers, so there is an interval when do not correspond.

Consider the new primary during a view change:

  1. Primary (`R₀`) receives DVC quorum for view `V₀`. `R₀` replaces its log. `R₀` begins repair.
  2. (... primary repairs)
  3. `R₀` completes repair. `R₀` transitions to `status=normal`. `R₀` bumps `view_normal`. `R₀` sends out `start_view` messages.

If another view change occurs during step 2 (with new primary `R₁`), `R₀` will send a DVC with a `view_normal` from _before_ `V₀`, but with the headers from `V₀`. But another replica in the view change may _also_ send a DVC, with the same `view_normal`, but the actual (different) headers!

### Not So Normal

Possibly `view_normal` should be renamed.
- For backups, "view normal" is accurate.
- But for primaries that completed completed their DVC quorum but did _not_ complete repair, it is inaccurate. They never reach normal status, but still need to advertise as if they did, to match their headers.

Maybe `view_active`? `view_write`? `view_member`?

This issue isn't present in VRR because VRR's DVCs include the entire log, and assume that it can be replace atomically.

sentientwaffle committed:
VSR: Messages prior to pipeline must be committed

The `commit_max` cannot fall behind `replica.op -| pipeline_max`.

Since DVCs include replicas' `commit_min` instead of `commit_max`, we can't just rely on taking the maximum from the DVCs.

sentientwaffle committed:
VSR/Correctness: Don't truncate canonical DVC

Summary

Truncating a canonical DVC's headers can fork the log.

VOPR Example

Consider a cluster with:

- Replica 2 is leading a view change (view=17).
- Replica 1 is participating in the view=17 view change.
- Replica 0 is partitioned.
- pipeline_prepare_queue_max = 6

The DVC quorum:

    2: on_do_view_change: replica=1 view_normal=15 op=38 commit_min=31
    2: on_do_view_change: replica=2 view_normal=14 op=33 commit_min=28

primary_set_log_from_do_view_change_messages():

  1. Replica 2's `op_canonical` is 28 (computed by primary_op_canonical_max()).
  2. Replica 1's DVC included headers 33,34,35,36,37,38. These were loaded into Replica 2's headers since Replica 1 has the canonical view.
  3. Near the end of primary_set_log_from_do_view_change_messages(), do_view_change_op_max() finds a hash break between op 32/33, and discards 33…38, forking the log.

It discarded the entire (canonical!) hash chain from replica 1! (In this case those ops were uncommitted, but that doesn't necessarily need to be true.) In any case, by discarding those ops we revert to our old (incorrect) op 29, which otherwise would have been repaired. We go on to commit the wrong op 29.

According to do_view_change_op_max(), the rationale for discarding those breaks (i.e. breaks which follow the primary's op_canonical) is:

    /// If there is a hash chain break, none of the headers from the canonical DVCs replaced
    /// the broken (leftover uncanonical) op.
    /// Removing these is necessary for correctness and liveness, to ensure that
    /// disconnected headers do not remain in place in lieu of gaps.

As the first sentence hints, passing more headers in the DVC would decrease the probability of this error, but that isn't an actual fix. (Aside: This bug was likely hidden until recently because it is becomes less probably for larger pipelines, and I have recently been heavily testing small pipelines.)

Simplified Example

With pipeline_prepare_queue_max=3:

  replica  headers                    commit_min  view_normal
        0  (1a 2a 3a 4a 5a 6a 7a 8a)          8a            4 (partitioned)
        1   1a 2a 3a 4a 5a{6a 7a 8a}          5a            4
        2   1a 2a 3a 4b                       3a            2

Prior to the do_view_change_op_max() truncation, replica's 2 headers are:

        2   1a 2a 3a 4b 5a 6a 7a 8a
                      ↑ hash break

Per CTRL, we can only discard headers that are *definitely uncommitted*.
A gap/break in replica 1's 6a,7a,8a would be definitively uncommitted and this safe to discard.
A break before _any_ of the view_normal_canonical's DVC headers though _must_ be repaired.

sentientwaffle committed:
VSR/Correctness: Bump primary view_normal early

Background

The new primary picks (in `primary_set_log_from_do_view_change_messages()`) at least one "canonical view". This is the highest `view_normal` from any DVC in its quorum.
The canonical DVC(s) headers are called the view's "canonical headers". These headers help select the new log head.

If multiple replicas have the same `view_normal`, there may be multiple canonical DVCs.
However, canonical headers _must_ be unambiguous. More specifically:

  - Any particular DVC may have gaps (1a, 2a, 4a), but no hash breaks (1a, 2a, 4b).
  - DVCs with the same `view_normal` _must_ not conflict.

Bug

*DVCs with the same `view_normal` _must_ not conflict.*

Currently they can conflict, though. The heart of this bug is that the `view_normal` updates at a different time than the headers, so there is an interval when do not correspond.

Consider the new primary during a view change:

  1. Primary (`R₀`) receives DVC quorum for view `V₀`. `R₀` replaces its log. `R₀` begins repair.
  2. (... primary repairs)
  3. `R₀` completes repair. `R₀` transitions to `status=normal`. `R₀` bumps `view_normal`. `R₀` sends out `start_view` messages.

If another view change occurs during step 2 (with new primary `R₁`), `R₀` will send a DVC with a `view_normal` from _before_ `V₀`, but with the headers from `V₀`. But another replica in the view change may _also_ send a DVC, with the same `view_normal`, but the actual (different) headers!

Not So Normal

Possibly `view_normal` should be renamed.
- For backups, "view normal" is accurate.
- But for primaries that completed completed their DVC quorum but did _not_ complete repair, it is inaccurate. They never reach normal status, but still need to advertise as if they did, to match their headers.

Maybe `view_active`? `view_write`? `view_member`?

This issue isn't present in VRR because VRR's DVCs include the entire log, and assume that it can be replace atomically.

jorangreef commented:
Great finds!!

> Possibly view_normal should be renamed.

I think that a better definition than "the last view for which the replica's status was normal” would be “the highest view that the replica participated in, or was witness to, i.e. received/sent a DVC or SV message”.

Therefore, `view_entered` or `view_joined` or `view_participated`. `view_member` is also good, although I think the first three put the emphasis more on the view that was participated in. Or we could do `view_change_seen`, because it's broader than participating in the view change, and also has to do with merely seeing the SV resulting from the view change.

sentientwaffle commented:
> “the highest view that the replica participated in, or was witness to, i.e. received/sent a DVC or SV message”.

:thinking: I don't think this is accurate.
- For the primary, `view_normal` must be bumped when it achieves a DVC quorum, since that is when it loads the new headers. This is after it has sent/received a DVC message, but before it sends the SV message.
- For the backup, `view_normal` must be bumped only when it receives the SV, not when it sends a DVC.

What about:

```
The highest view where
- the replica was a primary and acquired a DVC quorum, or
- the replica was a backup and received a SV message.
```

sentientwaffle committed:
VSR: Rename "view_normal" to "log_view"

The name "view_normal" is misleading now.

- For the primary, view_normal must be bumped when it achieves a DVC quorum, since that is when it loads the new headers. This is after it has sent/received a DVC message, but before it sends the SV message.
- For the backup, view_normal must be bumped only when it receives the SV, not when it sends a DVC.

The key difference between this implementation & VRR is:

- Our "logical" full log isn't the "real" full log, so the new primary's transition to normal status is deferred/asynchronous (and may never occur).
- Whereas in VRR the normal-status transition coincides exactly with the receipt of the new log. i.e. they are simultaneous.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1265,6 +1267,13 @@ pub fn ReplicaType(
             self.do_view_change_quorum = true;
 
             self.primary_set_log_from_do_view_change_messages();
+            // We aren't status=normal yet, but our headers from our prior log_view may have been
```
"from our prior log_view" reads nicely here.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3615,7 +3624,7 @@ pub fn ReplicaType(
 
         fn message_body_as_headers(message: *const Message) []const Header {
-            assert(message.header.size > @sizeOf(Header)); // Body must contain at least one header.
+            assert(message.header.size >= @sizeOf(Header)); // Body must contain at least one header.
```
I think this assert should stay as it was—the intention is that the *body* of the message, i.e. excluding the message header, should also contain at least one header.

Since this function only interprets the message body as headers, excluding the message header.

jorangreef replied:
In other words, we want `size` here to be > 128 bytes, since the 128 bytes here is only the message header, not anything in the body. If we allowed size to be `>= 128 bytes` then that would allow the body to have no headers in it.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3635,6 +3644,33 @@ pub fn ReplicaType(
+        /// - Given ops "2 3 4 5", return null.
+        fn do_view_change_headers_gap_min(message: *const Message) ?u64 {
+            assert(message.header.command == .do_view_change);
+            assert(message.header.size >= @sizeOf(Header));
```
This could be dropped since it's in `message_body_as_headers()` or otherwise should be corrected to be `>` not `>=` for the reasons explained above.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3635,6 +3644,33 @@ pub fn ReplicaType(
+
+            const headers = message_body_as_headers(message);
+            var op_gap: ?u64 = null;
+            var op = headers[0].op;
```
Since we make an assumption about which header's op to start with here, i.e. w.r.t. the ordering of headers, whether ascending or descending, it would be good to assert this order below.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3635,6 +3644,33 @@ pub fn ReplicaType(
+                op = header.op;
+                parent = header.parent;
+            }
+            return op_gap;
```
How dangerous could this return value be if we were to get it wrong? What extra safety checks could be done here?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5254,26 +5290,59 @@ pub fn ReplicaType(
+                        do_view_change_head.op -| constants.pipeline_max,
+                    );
+                    break :commit_max commit_max;
+                },
```
I think this could actually be cleaner using two nested inline `max()` calls, plus comma to layout. This avoids introducing the local variable, and avoids the need for a block.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5254,26 +5290,59 @@ pub fn ReplicaType(
 
-            // First, set all the canonical headers from the replica(s) with highest `view_normal`:
+            // "op_canonical" distinguishes between ops maybe-uncommitted ops from
+            // definitely-uncommitted ops — the latter are discarded to improve availablility.
```
```suggestion
            // definitely-uncommitted ops — the latter are discarded to improve availability.
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5254,26 +5290,59 @@ pub fn ReplicaType(
+            //
+            // Then replica 0 constructs:
+            //
+            //             1  2  3  4b        7  8  9
```
This is such a killer bug. Nice find again @sentientwaffle !

If I understand correctly, the bug is as simple (and dangerous!) as a primary that is fairly far behind, simply receiving a DVC with headers further along in the log, then truncating everything in between.

sentientwaffle replied:
Yeah, though the `7 8 9` must come from the canonical view, and not have been committed by that replica.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5254,26 +5290,59 @@ pub fn ReplicaType(
+            //
+            // If there is no gap in a DVC's headers, use its highest op — without a gap we must
+            // assume that all of the headers may be committed (at least until we run the nack
+            // protocol).
```
Great assumption to make, of course!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5287,7 +5356,12 @@ pub fn ReplicaType(
                     self.replace_headers(message_headers);
+                    op_canonical = std.math.max(
+                        op_canonical,
+                        do_view_change_headers_gap_min(message) orelse message_headers[0].op,
```
Could we have `do_view_change_headers_gap_min()` handle the `orelse` branch internally?

Alternatively, can we do more here to assert that the first header in `message_headers` is in fact the highest op?

sentientwaffle committed:
VSR: Canonical headers must agree

Fix an assertion failure.

When the new primary assembles the new view's headers, it:

1. Loads all canonical headers (`replace_header()`).
2. Uses the non-canonical headers to repair (`repair_header()`).
3. If there is a gap in the uncanonical headers, truncate it. (`do_view_change_op_max()` → `journal.remove_entries_from()`).
4. If there is a break in the new pipeline, truncate it. (`do_view_change_op_max()` → `journal.remove_entries_from()`).

But remember the DVC only includes `pipeline_max` headers. If we then truncate a break/gap, we may have as few as 1 header (`replica.op`) which is canonical, and the `pipeline_max - 1` headers preceding it may be from a previous view (and not part of the current view).
If we participate in another view change before we have had a chance to repair those last headers, our (potentially canonical!) DVC includes a hash break, which violates the DVC invariant.

I considered another approach where we just removed those discarded ops from the log, but that doesn't help when there is a gap `5a __ 7b`.

sentientwaffle committed:
VSR: Additional verification

sentientwaffle committed:
VSR: Fix DVC headers

`find_latest_headers_break_between()` requires that the headers on both ends exist in the journal's headers.
But in this case the `op_min` might be a gap or a break.

sentientwaffle committed:
Journal: stricter slot_with_header

We always called `slot_with_header()` after verifying `has(header)`, but `slot_with_header()` itself didn't check the checksum.
(I believe this is an artifact from the initial "slot_with"/"slot_for" API split).
Verifying the checksum better matches the usage and naming convention.

sentientwaffle committed:
Journal: Enforce invariants

- `find_latest_headers_break_between()`: I always forget the `op_min` must exist! I think there might actually be a bug in the replica around this, where we may need to fall back on `vsr_state.commit_min_checksum`.
- `write_prepare()`: Don't write a prepare if it is already being written. The `Replica` is already careful not to do this.

sentientwaffle committed:
Journal: Fix assertion failure

The Journal gets into a state where the header is clean, but `!prepare_inhabited`, which isn't supposed to be possible.

The problem is that a series of view changes lead to an op being rewritten several times quickly: version A₁, then version B, then version A₂ again.

`write_prepare_on_write_header()` ends up marking the entry (`A`) as clean (and the header is correct), but there is still a pending write (`write_prepare()`) to that same prepare (`A₂`) (with an interleaving `B`). That last `write_prepare` cleared the `prepare_inhabited`. Until the last prepare finishes, we are in a bad state where `prepare_inhabited` is false, but the entry is clean.

Seed: `10456`
Crash:

(Pay attention to `op=35`).

```
src/vsr/journal.zig:529:27: 0x369ec5 in vsr.journal.JournalType(vsr.replica.ReplicaType(state_machine.StateMachineType(test.storage.Storage,(struct state_machine.struct:25:70 constant)),test.message_bus.MessageBus,test.storage.Storage,test.time.Time),test.storage.Storage).has_clean (simulator)
                    assert(journal.prepare_inhabited[slot.index]);
                          ^
src/vsr/replica.zig:4887:39: 0x36a1b6 in vsr.replica.ReplicaType(state_machine.StateMachineType(test.storage.Storage,(struct state_machine.struct:25:70 constant)),test.message_bus.MessageBus,test.storage.Storage,test.time.Time).send_prepare_ok (simulator)
            if (self.journal.has_clean(header)) {
                                      ^
src/vsr/replica.zig:6082:33: 0x386f5b in vsr.replica.ReplicaType(state_machine.StateMachineType(test.storage.Storage,(struct state_machine.struct:25:70 constant)),test.message_bus.MessageBus,test.storage.Storage,test.time.Time).write_prepare_callback (simulator)
            self.send_prepare_ok(message.header);
                                ^
src/vsr/journal.zig:1781:27: 0x3c62db in vsr.journal.JournalType(vsr.replica.ReplicaType(state_machine.StateMachineType(test.storage.Storage,(struct state_machine.struct:25:70 constant)),test.message_bus.MessageBus,test.storage.Storage,test.time.Time),test.storage.Storage).write_prepare_release (simulator)
            write_callback(replica, wrote, write_trigger);
                          ^
src/vsr/journal.zig:1749:42: 0x3e2c5d in vsr.journal.JournalType(vsr.replica.ReplicaType(state_machine.StateMachineType(test.storage.Storage,(struct state_machine.struct:25:70 constant)),test.message_bus.MessageBus,test.storage.Storage,test.time.Time),test.storage.Storage).write_prepare_on_write_header (simulator)
            journal.write_prepare_release(write, message);
                                         ^
src/vsr/journal.zig:1895:27: 0x3dc87f in vsr.journal.JournalType(vsr.replica.ReplicaType(state_machine.StateMachineType(test.storage.Storage,(struct state_machine.struct:25:70 constant)),test.message_bus.MessageBus,test.storage.Storage,test.time.Time),test.storage.Storage).write_sectors_on_write (simulator)
            range.callback(write);
                          ^
src/test/storage.zig:418:23: 0x2ce6ac in test.storage.Storage.write_sectors_finish (simulator)
        write.callback(write);
                      ^
src/test/storage.zig:281:41: 0x2b22be in test.storage.Storage.tick (simulator)
            storage.write_sectors_finish(write);
                                        ^
src/simulator.zig:310:49: 0x2a5d40 in main (simulator)
                    cluster.storages[index].tick();
                                                ^
```

sentientwaffle committed:
VSR: More "primary_" renames

jamii commented on src/vsr/journal.zig:
```diff
@@ -586,6 +586,20 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+                if (header_prev.checksum != header_next.parent) return null;
+            } else return null;
+        }
+
```
Does this always return null?

sentientwaffle replied:
Good catch.. No, that's a bug. Though now I'm wondering why it made the test pass :thinking:

jamii replied:
Maybe the first two `return null`s should be `return op-1`?

jamii replied:
https://github.com/tigerbeetledb/tigerbeetle/issues/369#issuecomment-1360392012 :D

sentientwaffle replied:
> Maybe the first two return nulls should be return op-1?

Yeah, I think so too, but I need to figure out why it passes the tests without that

sentientwaffle replied:
Aha, I didn't check the final fix against the correct seed. The `return null` fails as expected after all, and `return op - 1` fixes it.

jamii reviewed (changes requested):
(Reviewable, ignore)

jamii reviewed (changes requested):
(Reviewable, ignore)

sentientwaffle committed:
Config: Check `journal_slot_count > pipeline_max`

And remove a duplicated assertion at the same time.

sentientwaffle committed:
VSR/Liveness: Include more headers in DVC

Summary

Each DVC includes headers, which are loaded by the new primary of a view change.
But with the intent to never send incorrect headers, DVCs were (in some cases) sending too few headers, such that the view change could not complete.

Example

Consider this DVC quorum (`pipeline_max=3`):

    replica   headers                       log_view
          0   1 [2  3  4b]                  4     (new primary)
          1   1  2  3  4a  5  6 [   8  9]   5     (old follower, incorrect!)
          2  (1  2  3  4a  5  6  7  8  9)   5     (old primary, partitioned)

Replica 0 would load the canonical headers from replica 1, resulting in a new log:

              1  2  3  4b           8  9

The on-DVC spots the gap at op 7 and concludes (correctly!) that ops 8/9 are uncommitted. They can (and must) be discarded.
But that leaves us without a head op! The new head op needs to be 6, since 5/6 are _definitely_ committed (based on `9+1 -| pipeline_max`).

Solution?

The intuitive solution is that replica 1's DVC should have included op 6:

          1   1  2  3  4a  5 [6     8  9]   5     (old follower, correct)

The replica 0 could truncate 8/9, without risking a fork.

But the solution is not as simple as "just include headers to the left of gaps" — we actually used to do that (pre-canonical-headers), but it has a critical flaw.
The old primary can wind up with a gap-break in its log suffix — that is, it could have a 6a,-,7b sequence where the gap "conceals" a chain break.
(It happens like this:
1. When the old view is started, the old primary truncated some ops after a gap.
2. Before the old primary could repair its log, another view change is triggered.
3. By truncating its log, the window of the last "pipeline_max" ops "moves" backwards, potentially to include a break or gap-break.)

Solution.

The key insight it to consider (and implement) these two cases separately:

- If the replica was a primary in its last log_view, then we can't hop over gaps — but we also don't need to!
  We know that the consecutive suffix is sufficient to start the new view, because if it wasn't, the old primary couldn't have finished its DVC.
  (If the old primary *did* finish its DVC, then it would have a pristine log — fully repaired, no gaps or breaks in the suffix.)
- If the replica was a follower in its last log_view, then any gaps in the pipeline-max-suffix *must* be due to missed prepares
  (e.g. a follower that receives op 5, then op 7, then op 9).
  So it is safe (within the pipeline-max-suffix) to hop over gaps and include the header on the left side.
  (A follower always starts a new view by receiving a SV message, which includes at least pipeline_max consecutive ops).

There is one known bug in this commit (fix in the next) because I couldn't untangle the diff to include it here:
In `primary_set_log_from_do_view_change_messages()`, `op_canonical` is set to `gap` when it should be set to `gap - 1`.

sentientwaffle committed:
VSR: Refactor DVC, fix gap-1 issue

In `primary_set_log_from_do_view_change_messages()`, the `op_canonical` should be `gap - 1`, not `gap` (i.e. the canonical op must be an available header, not the gap itself).

Refactor the DVC logic to reduce its statefulness — move much of the logic into pure functions (grouped in the `DVCQuorum` namespace).

The major change here is how canonical headers are loaded:
Instead of looping all of the DVC messages and filtering by canonical and using `replace_header()`, use an iterator that runs over all canonical DVCs simultaneously.
This enables some extra verification, since canonical DVCs must never disagree.

Add `message_body_as_headers_chain_consecutive()` and `message_body_as_headers_chain_disjoint()`, which are just `message_body_as_headers()` with extra verification.

sentientwaffle committed:
VSR: Simplify primary_op_canonical_max()

sentientwaffle committed:
VSR: Massively simplify view installation

Previously view installation (at a high level):

1. Set the new head op, truncating the log.
2. Replace the new primary's log entries with canonical headers from each canonical DVC.
3. Repair the new primary's log entries with non-canonical headers from each non-canonical DVC.
4. Comb through the log's suffix for breaks/gaps that indicate uncommitted ops (`primary_do_view_change_op_max()`).
5. Set the new head op, truncating the log (again).

This can be simplified:
- The "comb through the log's suffix" step is extremely complicated, but this is avoidable.
  Thanks to the DVC invariants, we can avoid any hash breaks by searching for gaps _prior_ to installing any new entries to the log.
- Instead of setting the head op & truncating the log twice, just do it once.

The core DVC logic is extracted into `headers_canonical()`, which is a pure function over only the DVC quorum messages.

sentientwaffle committed:
VSR: DVCQuorum.verify()

jamii commented on src/vsr/replica.zig:
```diff
@@ -138,13 +138,16 @@ pub fn ReplicaType(
+        /// The latest view where
+        /// - the replica was a primary and acquired a DVC quorum, or
+        /// - the replica was a backup and received a SV message.
+        log_view: u32,
```
Can we add why we care about this. I think this is correct:

```suggestion
        /// ie the latest view for which we can guarantee that this replica was not missing any committed ops.
        log_view: u32,
```

jamii replied:
I guess that's not quite correct - if it was the primary it might set log_view before completing repair?

sentientwaffle replied:
Right, it should be more along the lines of:

> ie the latest view in which this replica changed its head message

The primary of a new view updates its log_view before completing repair, and followers joining a view (via `start_view`) update their `log_view` immediately, and then repair asynchronously.

jamii commented:
bors d+

I'm not confident I would spot any logic bugs, but I didn't find any more mechanical bugs and I'm fairly confident that the changes to the code and comments agree.

sentientwaffle committed:
VSR: Document log_view

sentientwaffle commented:
bors r+

sentientwaffle commented:
bors r+

Retrying again, with https://github.com/tigerbeetledb/tigerbeetle/pull/392

sentientwaffle commented:
bors retry

Failed due to concurrent PR, trying again...

sentientwaffle commented:
bors r+
